File: 2020/day08/python/day82.py
def main():
    opss = make_all_ops_combinations()
    for ops in opss:
        run_program(ops)

def run_program(ops):
    pc = 0
    acc = 0
    while(True):
        if pc >= len(ops):
            print("Finished!", acc)
            break
        if pc < 0: break
        visited, op_name, val =ops[pc]
        if visited : break
        
        ops[pc][0] = True
        pc += 1
        if op_name == "nop": continue
        if op_name == "acc": acc += int(val)
        if op_name == "jmp": 
            pc += int(val)-1
            continue
    return acc

def make_all_ops_combinations():
    ops = [line.strip() for line in open("../input.txt")]
    ops = [[False, line[:3], line[3:]] for line in ops]
    opss = [ops]
    for idx in range(len(ops)):
        new_ops = [x[:] for x in ops]
        curr_op = ops[idx]
        _, op_name, _ = curr_op
        if op_name == "acc": continue
        if op_name == "jmp": new_ops[idx][1] = "nop"
        if op_name == "nop": new_ops[idx][1] = "jmp"
        opss.append(new_ops)
    return opss

if __name__ == "__main__":
    main()
    

File: 2020/day08/haskell/day82.hs
import Data.Sequence (fromList, update, index, Seq)
main = interact $ show 
    . maximum
    . (\opss -> map (\ops -> runProgram 0 0 ops) opss)
    . (\seq -> map (\x -> genOps seq x)  [0..(length seq-1)]) 
    . fromList . zip (repeat False) . map parse . lines

runProgram :: Int -> Int -> Seq (Bool, (String, Int)) -> Int
runProgram acc n list = let (visited,(op_name, num)) = list `index` n in
    if n >= length list then acc 
    else if n < 0 then 0 
    else if visited then 0
        else let newList = update n (True, (op_name, num)) list in 
            case op_name of 
                "nop" -> runProgram acc (n+1) newList 
                "acc" -> runProgram (acc+num) (n+1) newList
                "jmp" -> runProgram acc (n+num) newList 

parse :: String -> (String, Int)
parse = (\(op:snum:xs) -> (op, parseNum snum)) . words

parseNum :: String -> Int
parseNum ('+':num) = read num
parseNum ('-':num) = negate $ read num

genOps :: Seq (Bool, (String, Int)) -> Int -> Seq (Bool, (String, Int))
genOps list n = let (visited,(op_name, num)) = list `index` n in
    case op_name of
        "acc" -> list
        "nop" -> update n (visited, ("jmp", num)) list
        "jmp" -> update n (visited, ("nop", num)) list